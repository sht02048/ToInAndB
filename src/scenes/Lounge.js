import Renderer from "../graphics/Renderer";
import Cross from "../entities/enemies/Cross";
import Wings from "../entities/enemies/Wings";

class Lounge extends Renderer {
  #isDone = true;

  constructor() {
    super();

    this.wings = new Wings({ x: this.canvasWidth - 200, y: -100 });
    this.crossList = [
      new Cross({ x: 100, y: -100 }),
      new Cross({ x: 600, y: -150 }),
    ];

    this.shouldBeDisplayed = false;
    this.hasStarted = false;
  }

  update() {
    this.frame += 1;
    this.respawnCross();

    this.wings.update();
    this.crossList.forEach((cross) => cross.update());
  }

  render() {
    this.wings.render();
    this.crossList.forEach((cross) => cross.render());
  }

  setTarget(player) {
    this.wings.setTargetList(player);
    this.crossList.forEach((cross) => {
      cross.setTargetList(player);
    });

    this.player = player;
  }

  setSceneTargetList() {
    const sceneTargetList = [this.wings];

    this.crossList.forEach((cross) => sceneTargetList.push(cross));

    return sceneTargetList;
  }

  checkSceneStatus() {
    this.#isDone = true;

    if (!this.wings.isDestroyed && !this.wings.isVanished) {
      this.#isDone = false;
    }

    this.crossList.forEach((cross) => {
      if (!cross.isDestroyed && !cross.isVanished) {
        this.#isDone = false;
      }
    });

    return this.#isDone;
  }

  respawnCross() {
    if (this.frame % 300 !== 0 || this.wings.isDestroyed) {
      return;
    }

    const firstRandomX =
      Math.random() * (this.canvasWidth / 2 - this.minX) + this.minX;
    const secondRandomX =
      Math.random() * (this.maxX - this.canvasWidth / 2) + this.canvasWidth / 2;
    const firstRandomY = Math.random() * 100 + 44;
    const secondRandomY = Math.random() * 100 + 44;

    const firstCross = new Cross({ x: firstRandomX, y: -firstRandomY });
    const secondCross = new Cross({ x: secondRandomX, y: -secondRandomY });

    firstCross.setTargetList(this.player);
    secondCross.setTargetList(this.player);

    this.crossList.push(firstCross, secondCross);
  }
}

export default Lounge;
