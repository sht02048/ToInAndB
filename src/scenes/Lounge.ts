import Scene from "./Scene";

import Player from "../entities/Player";
import Cross from "../entities/enemies/Cross";
import Wings from "../entities/enemies/Wings";

class Lounge extends Scene {
  private wings: Wings;
  private crossList: Cross[];
  private player: Player;

  constructor() {
    super();

    this.wings = new Wings({ x: this.canvasWidth - 200, y: -100 });
    this.crossList = [
      new Cross({ x: 100, y: -100 }),
      new Cross({ x: 600, y: -150 }),
    ];
  }

  update() {
    this.frame += 1;
    this.respawnCross();

    this.wings.update();
    this.crossList.forEach((cross) => cross.update());
  }

  render() {
    this.wings.render();
    this.crossList.forEach((cross) => cross.render());
  }

  setTarget(player: Player) {
    this.wings.setTargetList([player]);
    this.crossList.forEach((cross) => {
      cross.setTargetList([player]);
    });

    this.player = player;
  }

  setSceneTargetList() {
    const sceneTargetList = [this.wings, ...this.crossList];

    return sceneTargetList;
  }

  checkSceneStatus() {
    this.isDone = true;

    if (!this.wings.isDestroyed && !this.wings.isVanished) {
      this.isDone = false;
    }

    this.crossList.forEach((cross) => {
      if (!cross.isDestroyed && !cross.isVanished) {
        this.isDone = false;
      }
    });

    return this.isDone;
  }

  respawnCross() {
    if (this.frame % 300 !== 0 || this.wings.isDestroyed) {
      return;
    }

    const firstRandomX =
      Math.random() * (this.canvasWidth / 2 - this.minX) + this.minX;
    const secondRandomX =
      Math.random() * (this.maxX - this.canvasWidth / 2) + this.canvasWidth / 2;
    const firstRandomY = Math.random() * 100 + 44;
    const secondRandomY = Math.random() * 100 + 44;

    const firstCross: Cross = new Cross({ x: firstRandomX, y: -firstRandomY });
    const secondCross: Cross = new Cross({
      x: secondRandomX,
      y: -secondRandomY,
    });

    firstCross.setTargetList(this.player);
    secondCross.setTargetList(this.player);

    this.crossList.push(firstCross, secondCross);
  }
}

export default Lounge;
