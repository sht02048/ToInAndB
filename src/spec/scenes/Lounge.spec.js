import Lounge from "../../scenes/Lounge";

describe("Lounge", () => {
  const mainCanvas = document.createElement("canvas");
  const introCanvas = document.createElement("canvas");
  mainCanvas.id = "main-canvas";
  introCanvas.id = "intro-canvas";

  document.body.appendChild(mainCanvas);
  document.body.appendChild(introCanvas);

  let lounge;

  beforeEach(() => {
    lounge = new Lounge();
  });

  it("should have proper functions", () => {
    expect(lounge.update).to.be.a("function");
    expect(lounge.render).to.be.a("function");
    expect(lounge.setTarget).to.be.a("function");
    expect(lounge.setSceneTargetList).to.be.a("function");
    expect(lounge.checkSceneStatus).to.be.a("function");
  });

  it("should render wings and cross", () => {
    lounge.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should respawn cross", () => {
    const originalCrossNumber = lounge.crossList.length;
    lounge.frame = 299;
    lounge.update();

    expect(originalCrossNumber).to.not.equal(lounge.crossList.length);
  });

  it("should check if scene is done", () => {
    lounge.wings.isDestroyed = true;
    lounge.crossList.forEach((cross) => (cross.isDestroyed = true));

    expect(lounge.checkSceneStatus()).to.equal(true);
  });
});
