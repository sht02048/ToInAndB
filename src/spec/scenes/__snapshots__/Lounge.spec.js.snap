// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lounge > should render wings and cross 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 100,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/wings.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 100,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/cross.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 600,
      "dy": -150,
      "img": <img
        src="/src/assets/sprites/enemies/cross.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
